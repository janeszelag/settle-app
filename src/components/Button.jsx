import React from 'react'
import { useWindowDimensions, Pressable, StyleSheet, Text } from 'react-native'
import { gunmetal, white, cedarChest } from '../styles/colors'


const Button = ({ children, variant, onPress, mode }) => {
    const windowWidth = useWindowDimensions().width
    const findWidth = () => {
        return windowWidth <= 768 ? { width: '100%' } : { width: 180 }
    }
    let textColor

    const getVariant = (variant, mode) => {
        if (variant === 'outline') {
            textColor = mode === 'dark' ? gunmetal : cedarChest
            return {
                backgroundColor: white,
                borderColor: textColor,
                borderWidth: 1,
            }
        } else {
            textColor = white
            return {
                backgroundColor: mode === 'dark' ? gunmetal : cedarChest
            }
        }
    }
    return (
        <Pressable
            onPress={onPress}
            style={[styles.wrapperCustom, findWidth(), getVariant(variant, mode)]}>
            <Text style={[styles.text, { color: textColor }]}>{children}</Text>
        </Pressable>
    )

}

const styles = StyleSheet.create({
    wrapperCustom: {
        borderRadius: 4,
        padding: 6,
        height: 48,
        justifyContent: 'center',
        alignItems: 'center',
        //cursor: 'pointer',
    },
    text: {
        fontSize: 16,
    },

});

export default Button

