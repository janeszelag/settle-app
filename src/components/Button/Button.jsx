import React, { useState, useRef } from 'react'
import { useWindowDimensions, Pressable, StyleSheet, Text, Platform } from 'react-native'
import PropTypes from 'prop-types'
import {
  gunmetal,
  white,
  cedarChest,
  lightCedarChest,
  lightGunmetal,
  darkCedarChest
} from '../../styles/colors'
import { useHover, useFocus, useActive } from 'react-native-web-hooks'

const Button = ({ children, variant, onPress, colour }) => {

  const ref = useRef(null);
  const [isPressed, setIsPressed] = useState(false)
  const windowWidth = useWindowDimensions().width
  const isFocused = useFocus(ref)
  const isHover = useHover(ref)
  const isActive = useActive(ref)


  const pressBackground = colour === 'dark' ? lightGunmetal : lightCedarChest
  const modeColor = colour === 'dark' ? gunmetal : cedarChest

  const findWidth = () => {
    return windowWidth <= 768 ? { width: '100%' } : { width: 180 }
  }

  const getTextStyles = (variant, colour) => {
    const textColor = colour === 'dark' ? gunmetal : cedarChest
    const pressColor = colour === 'dark' ? gunmetal : darkCedarChest
    if (Platform.OS === 'web') {
      return {
        color: variant === 'outline' ?
          (isPressed ? pressColor : textColor) :
          (isPressed || isHover ? textColor : white)
      }
    } else {
      return {
        color: variant === 'outline' ?
          (isPressed ? pressColor : textColor) :
          white
      }
    }

  }

  const getWebStyles = (variant) => {
    if (Platform.OS !== 'web') return
    if (variant === 'outline') {
      return {
        cursor: 'pointer',
        boxShadow: isFocused && `0 0 0 0, 0 0 0 0.125rem ${white} inset, 0 0 0 0.1875rem ${modeColor} inset`,
        transition: 'all 0.2s ease-in-out',
        borderWidth: isHover || isFocused ? '0.125rem' : 1
      }
    } else {
      return {
        cursor: 'pointer',
        backgroundColor: isHover || isPressed ? (isActive ? pressBackground : white) : modeColor,
        borderColor: modeColor,
        borderWidth: isHover || isFocused ? '0.125rem' : 1,
        color: isHover || isPressed ? modeColor : white
      }
    }
  }

  const getVariant = (variant, colour) => {
    if (variant === 'outline') {
      return {
        backgroundColor: isPressed ? pressBackground : white,
        borderColor: modeColor,
        borderWidth: isPressed ? 1.5 : 1,
      }
    } else {
      return {
        backgroundColor: colour === 'dark' ? gunmetal : cedarChest,
        opacity: isPressed && Platform.OS !== 'web' ? 0.85 : 1
      }
    }

  }

  return (
    <Pressable
      onPress={onPress}
      onPressOut={() => setIsPressed(false)}
      onPressIn={() => setIsPressed(true)}
      ref={ref}
      accessibilityRole="button"
      style={[styles.wrapperCustom, findWidth(), getVariant(variant, colour), getWebStyles(variant)]}>
      <Text style={[styles.text, getTextStyles(variant, colour)]}>{children}</Text>
    </Pressable>
  )
}

const styles = StyleSheet.create({
  wrapperCustom: {
    borderRadius: 6,
    padding: 6,
    height: 48,
    justifyContent: 'center',
    alignItems: 'center',
  },
  text: {
    fontSize: 16,
  },
})

Button.propTypes = {
  /**
   * The style.
   */
  variant: PropTypes.oneOf(['outline', 'filled']),
  /**
   * The colour.
   */
  colour: PropTypes.oneOf(['dark', 'orange']),
  /**
   * The label.
   */
  children: PropTypes.string.isRequired,
}
Button.defaultProps = {
  variant: 'filled',
  colour: 'orange'
}

export default Button

