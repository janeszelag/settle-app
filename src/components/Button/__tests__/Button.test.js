import React from "react"
import renderer from "react-test-renderer"
import Button from "../Button"

describe("<Button />", () => {
  it("has 1 child", () => {
    const tree = renderer.create(<Button>Words</Button>).toJSON()
    expect(tree.children.length).toBe(1)
  })

  it("renders", () => {
    const tree = renderer
      .create(<Button>Text Button</Button>)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("renders an outline variant with the default colour as 'orange'", () => {
    const tree = renderer
      .create(<Button variant='outline'>Text Button</Button>)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("renders an outline variant in gunmetal", () => {
    const tree = renderer
      .create(
        <Button variant='outline' colour='dark'>
          Text Button
        </Button>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("renders a filled variant with the default colour as 'orange'", () => {
    const tree = renderer
      .create(<Button variant='filled'>Text Button</Button>)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("renders a filled variant in gunmetal", () => {
    const tree = renderer
      .create(<Button variant='filled'>Text Button</Button>)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
