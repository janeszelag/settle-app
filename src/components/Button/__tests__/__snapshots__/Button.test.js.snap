// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders 1`] = `
<View
  accessibilityRole="button"
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 6,
        "height": 48,
        "justifyContent": "center",
        "padding": 6,
      },
      Object {
        "width": "100%",
      },
      Object {
        "backgroundColor": "#c8553d",
        "opacity": 1,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 16,
        },
        Object {
          "color": "#ffffff",
        },
      ]
    }
  >
    Text Button
  </Text>
</View>
`;

exports[`<Button /> renders a filled variant in gunmetal 1`] = `
<View
  accessibilityRole="button"
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 6,
        "height": 48,
        "justifyContent": "center",
        "padding": 6,
      },
      Object {
        "width": "100%",
      },
      Object {
        "backgroundColor": "#c8553d",
        "opacity": 1,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 16,
        },
        Object {
          "color": "#ffffff",
        },
      ]
    }
  >
    Text Button
  </Text>
</View>
`;

exports[`<Button /> renders a filled variant with the default colour as 'orange' 1`] = `
<View
  accessibilityRole="button"
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 6,
        "height": 48,
        "justifyContent": "center",
        "padding": 6,
      },
      Object {
        "width": "100%",
      },
      Object {
        "backgroundColor": "#c8553d",
        "opacity": 1,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 16,
        },
        Object {
          "color": "#ffffff",
        },
      ]
    }
  >
    Text Button
  </Text>
</View>
`;

exports[`<Button /> renders an outline variant in gunmetal 1`] = `
<View
  accessibilityRole="button"
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 6,
        "height": 48,
        "justifyContent": "center",
        "padding": 6,
      },
      Object {
        "width": "100%",
      },
      Object {
        "backgroundColor": "#ffffff",
        "borderColor": "#293241",
        "borderWidth": 1,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 16,
        },
        Object {
          "color": "#293241",
        },
      ]
    }
  >
    Text Button
  </Text>
</View>
`;

exports[`<Button /> renders an outline variant with the default colour as 'orange' 1`] = `
<View
  accessibilityRole="button"
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 6,
        "height": 48,
        "justifyContent": "center",
        "padding": 6,
      },
      Object {
        "width": "100%",
      },
      Object {
        "backgroundColor": "#ffffff",
        "borderColor": "#c8553d",
        "borderWidth": 1,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 16,
        },
        Object {
          "color": "#c8553d",
        },
      ]
    }
  >
    Text Button
  </Text>
</View>
`;
